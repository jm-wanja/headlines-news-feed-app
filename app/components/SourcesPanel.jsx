import React from 'react';
import Select from 'react-select';
import PropTypes from 'prop-types';
import { getNewsArticles } from '../actions/AppActions';
import SourcesStore from '../stores/SourcesStore';

/**
 * Class  displaying the Search Form.
 * @extends React.Component
 */
class SourcesPanel extends React.Component {
  /**
   * Returns the value in the Search Field
   * @param {object} props - The properties of the News Sources Class
   */
  constructor(props) {
    super(props);
    this.state = {
      currentValue: '',
      clearable: false,
    };
    this.mapStateToOptions = this.mapStateToOptions.bind(this);
    this.updateSearch = this.updateSearch.bind(this);
  }

  /**
   * Method to set the currently selected news source,
   * Method to send request to the App Actions.
   * @param {event} event - the select-box change event
   * @return {*} return the new state and props
   */
  updateSearch(event) {
    const value = event.value;
    this.props.setIsLoading(true);
    this.setState({
      currentValue: value,
      newsSource: `Viewing news from ${event.label}`,
      sourceDescription: event.description,
    });
    // console.log('getNewsArticles', value)
    // then action getnewsarticles is dispatched
    getNewsArticles(value);
    // const sortBy = value.split('?sortBy=');
    // this.props.setSortBy(sortBy);
  }

  /**
   * Method to generate the options for the Search box.
   * @param {sources} sources - an array of all the news sources
   * @return {object} return each news source
   */
  mapStateToOptions(sources) { // function that returns an object with options for my form = all sources
    this.sourcesMap = sources;
    return sources.map(source => {
      return {
        value: `${source.id}`,
        label: source.name,
        description: source.description,
      }
    });
  }

  /**
   * Renders the Search Input
   * @return {*} render the Search Form
   */
  render() {

    // const sourcesComp = this.props.sources.map(source =>
    //   <Sources key={source.id} data={source} />
    // )

    return (

      <div className="">
        <div className="section no-pad-bot">
          <div className="container">
            <h4 className="header center pink-text">Know what's happening now</h4>
            <div className="row center">
              <h5 className="header col s12 light">
                Search through our current set of about 70 news sources!!!
              </h5>
            </div>
            <div className="row center valign-wrapper">
              <div className="search-box col s12 m6 offset-m3">
                <Select
                  name="form-field-name"
                  options={this.mapStateToOptions(this.props.sources)}
                  value={this.state.currentValue}
                  className="search-bar"
                  onChange={this.updateSearch} // on change handler 
                  autofocus
                  clearable={this.state.clearable}
                  placeholder="Select News Source" // displayed when there's no value
                />
              </div>
            </div>
            <h5 className="center">{this.state.newsSource}</h5>
            <p className="center light"> {this.state.sourceDescription}</p>
          </div>
        </div>
        {/*<div> It should render here </div>
        <DefaultPage sources={this.props.sources} />*/}
      </div>
    );
  }
}

/**
 * Set the PropTypes for Sources Panel
 */
SourcesPanel.propTypes = {
  sources: PropTypes.array.isRequired,
  setSortBy: PropTypes.func.isRequired,
  setIsLoading: PropTypes.func,
};

export default SourcesPanel;
