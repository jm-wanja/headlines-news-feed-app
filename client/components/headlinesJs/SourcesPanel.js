import React from 'react';
export default class SourcesPanel extends React.Component {
    render() {
        const data = this.props.data;
        return (
            <div>
                {data.length && data.map((dataObject) => (
                    <div> {dataObject.name} </div>
                ))}
            </div>
        );
    }
}
