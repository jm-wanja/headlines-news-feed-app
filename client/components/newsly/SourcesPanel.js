import React from 'react';
import request from 'superagent';
import { Layout, Drawer, Navigation } from "react-mdl";

export default class SourcesPanel extends React.Component {
    constructor(props) {
        super(props);//def this.props 
        this.state = { //initialize state because the data will change
            sources: [] //empty array because the data from api is in an array
        }
    }

    componentWillMount() { //an instance of the component created
        this.fetchSources().then((data) => {
            this.setState({ //call setstate to change state to populate data into it
                sources: data.body.sources
            });
        });
    }
    fetchSources() {
        return (
            request
                .get('https://newsapi.org/v1/sources?language=en')
        )
    }
    render() {
        console.log("State", this.state);
        return (
            <Drawer title="SOURCES">
                <Navigation>
                    {this.state.sources.length ? this.state.sources.map((source) => (
                        <div id="sources" key={source.id}>
                            <a id="sources-link" href={source.url} target="_blank">{source.name}</a>
                        </div>
                    )) :
                        <div id="loading">Loading...</div>
                    }
                </Navigation>
            </Drawer>
        );
    }
}
