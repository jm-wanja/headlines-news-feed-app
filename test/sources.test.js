import React from 'react';
import { mount, shallow } from 'enzyme';
import shallowToJSON from 'enzyme-to-json';
import SourcesPanel from '../client/components/headlinesJs/sourcesPanel.js';

// describe('App', () => {
//   it('renders Headlines App Setup', () => {
//     const headlines = shallow(<App />);
//     expect(headlines.find('div').text()).toEqual('Headlines App Setup');
//   });
// });

// describe('Headlines', () => {
//   it('renders sourcesPanel', () => {
//     const sources = shallow(<Headlines />);
//     expect(sources.find('.sourcesPanel').toBe.defined);
//   }); 
// })

describe('Sources Panel Component', () => {
  it('snapshot of SourcePanel', () => {
    const sources = shallow(<SourcesPanel data={ {} }/>);
    const tree = shallowToJSON(sources);
    expect(tree).toMatchSnapshot();
  }); 
})