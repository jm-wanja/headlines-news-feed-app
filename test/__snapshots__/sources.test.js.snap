// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sources Panel Component snapshot of SourcePanel 1`] = `<div />`;
